pkg/service/configcache/service.go

```diff
@@ -0,0 +1,217 @@
+// Use svc.Run() to let the cache update itself periodically with the current configuration.
+//
+// Current implementation doesn't care about changing IP, what means that the configuration for given cluster and host
+// will stay there forever.
+type Service struct {
+	clusterSvc   cluster.Servicer
+	scyllaClient scyllaclient.ProviderFunc
+	secretsStore store.Store
+
+	configs sync.Map
+	logger  log.Logger
+}
+
+// NewService is the constructor for the cluster config cache service.
+func NewService(clusterSvc cluster.Servicer, client scyllaclient.ProviderFunc, secretsStore store.Store, logger log.Logger) ConfigCacher {
+	return &Service{
+		clusterSvc:   clusterSvc,
+		scyllaClient: client,
+		secretsStore: secretsStore,
+		configs:      sync.Map{},
+		logger:       logger,
+	}
+}
+
+// Init updates cache with config of all currently managed clusters.
+func (svc *Service) Init(ctx context.Context) {
+	svc.updateAll(ctx)
+}
+
+// Read returns either the host configuration that is currently stored in the cache,
+// ErrNoClusterConfig if config for the whole cluster doesn't exist,
+// or ErrNoHostConfig if config of the particular host doesn't exist.
+func (svc *Service) Read(clusterID uuid.UUID, host string) (NodeConfig, error) {
+	emptyConfig := NodeConfig{}
+
+	clusterConfig, err := svc.readClusterConfig(clusterID)
+	if err != nil {
+		return emptyConfig, err
+	}
+
+	rawHostConfig, ok := clusterConfig.Load(host)
+	if !ok {
+		return emptyConfig, ErrNoHostConfig
+	}
+	hostConfig, ok := rawHostConfig.(NodeConfig)
+	if !ok {
+		panic("cluster host emptyConfig cache stores unexpected type")
+	}
+	return hostConfig, nil
+}
+
+// Run starts the infinity loop responsible for updating the clusters configuration periodically.
+func (svc *Service) Run(ctx context.Context, skipFirst bool) {
+	if skipFirst {
+		time.Sleep(updateFrequency)
+	}
+	freq := time.NewTicker(updateFrequency)
+
+	for {
+		// make sure to shut down when the context is cancelled
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-freq.C:
+			svc.updateAll(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// ForceUpdateCluster updates single cluster config in cache and does it outside the background process.
+// There is no synchronization at all. Synchronization as guaranteed by the sync.Map used to back the cache.
+// If the ForceUpdateCluster is executed at the same time when the cluster config is being updated by the
+// background process, then the configuration will be just updated twice.
+func (svc *Service) ForceUpdateCluster(ctx context.Context, clusterID uuid.UUID) bool {
+	logger := svc.logger.Named("Force update cluster").With("cluster", clusterID)
+
+	c, err := svc.clusterSvc.GetCluster(ctx, clusterID.String())
+	if err != nil {
+		logger.Error(ctx, "Update failed", "error", err)
+	}
+
+	return svc.updateSingle(ctx, c)
+}
+
+func (svc *Service) updateSingle(ctx context.Context, c *cluster.Cluster) bool {
```

Comment:
ohh, ForceUpdateCluster in interface returns bool as well, let's leave it
it's additional information that can be used, but doesn't have to be used

Reply:
But what does this bool mean? It can be set to true even when we failed to fetch all nodes info? For me it's more confusing than useful (even in the future). I guess that changing it to an error (which should be checked) returned when client cannot be created makes more sense.

Reply:
errors are logged in the method body
let's leave it
I don't see it as a big deal